### PR TITLE
CASSANDRA-16806. Allow TRUNCATE and DELETE to work on Virtual Tables if the implementation allows it.

### PR DESCRIPTION
**Pull request by azotcsit**
Summary of changes:
- added `truncate()` method to `VirtualTable` interface
- made necessary changes to handle virtual tables in `TruncateStatement`
- added `AbstractWritableVirtualTable` with necessary methods for granular deletes and updates
- added a restriction for custom timestamps
- updated unit tests
- updated docs and CHANGES.txt

### CODE REVIEW

File: doc/source/new/virtualtables.rst [hunk 1]
```diff
@@ -42,11 +42,12 @@ Virtual tables and virtual keyspaces are quite different from regular tables and
 - Virtual tables are not replicated
 - Virtual tables are local only and non distributed
 - Virtual tables have no associated SSTables
+- Virtual tables support ``TRUNCATE`` only if the underlaying implementation allows it
 - Consistency level of the queries sent virtual tables are ignored
-- Virtual tables are managed by Cassandra and a user cannot run  DDL to create new virtual tables or DML to modify existing virtual       tables
```
Comment (azotcsit) [inline doc/source/new/virtualtables.rst:46]: ``TRUNCATE`` is a DML

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 2]
```diff
@@ -88,17 +89,45 @@ public void apply(PartitionUpdate update)
...
+        {
+            isTruncated = true;
+        }
+    }
+
     @BeforeClass
     public static void setUpClass()
     {
-        TableMetadata vt1Metadata =
-        TableMetadata.builder(KS_NAME, VT1_NAME)
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:95]: It was badly formatted.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 3]
```diff
@@ -118,16 +147,15 @@ public DataSet data()
         };
         VirtualTable vt2 = new WritableVirtualTable(KS_NAME, VT2_NAME);
 
-        TableMetadata vt3Metadata =
-        TableMetadata.builder(KS_NAME, VT3_NAME)
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:122]: It was badly formatted.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 4]
```diff
@@ -88,17 +89,45 @@ public void apply(PartitionUpdate update)
...
+
+        TruncatableVirtualTable(TableMetadata metadata)
+        {
+            super(metadata);
+        }
+
+        @Override
+        public DataSet data()
+        {
+            if (isTruncated)
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: Just a simple way to emulate truncation. Theoretically I could use a collection, smth like that:
```
TruncatableVirtualTable(TableMetadata metadata, Map<String, Integer> refToCleanupData) 

public void truncate()
{
        refToCleanupData.clear();
}
```
```
Map<String, Integer> truncatableData = <some data>;
VirtualTable vt4 = new TruncatableVirtualTable(vt4Metadata, truncatableData)
{

    public DataSet data()
    {
            SimpleDataSet data = new SimpleDataSet(metadata());
            truncatableData.forEach((key, value) -> data.row(key).column("value", value));
            return data;
    }

}
```

This way looks more similar to a real use case. Please, let me know if this one is more preferable for a unit test.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 5]
```diff
@@ -317,6 +365,32 @@ public void testModifications() throws Throwable
...
+
+        // truncate
+        execute("TRUNCATE test_virtual_ks.vt4");
+
+        // check that truncation is applied
+        assertEmpty(execute("SELECT * FROM test_virtual_ks.vt4"));
+    }
+
+    @Test
+    public void testInvalidDMLOperations() throws Throwable
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: I moved truncation test from `testInvalidDDLOperations` to `testInvalidDMLOperations` because `TRUNCATE` is a DML operation. Also I added a test for `INSERT` into read-only VT - it was missing.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 6]
```diff
@@ -317,6 +365,32 @@ public void testModifications() throws Throwable
                              "INSERT INTO test_virtual_ks.vt2 (key, value) VALUES ('pk2', 22) IF NOT EXISTS");
     }
 
+    @Test
+    public void testTruncationOnTruncatableTable() throws Throwable
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: This is a unit test to cover the actual change.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 7]
```diff
@@ -317,6 +365,32 @@ public void testModifications() throws Throwable
...
+        execute("TRUNCATE test_virtual_ks.vt4");
+
+        // check that truncation is applied
+        assertEmpty(execute("SELECT * FROM test_virtual_ks.vt4"));
+    }
+
+    @Test
+    public void testInvalidDMLOperations() throws Throwable
+    {
+        assertInvalidMessage("Modification is not supported by table test_virtual_ks.vt1",
```
Comment (alex-ninja) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:386]: This use case was not covered, so I added such a test.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java (hunk 4 above)
Comment (blerer) [reply test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: Checking that perfoming a TRUNCATE call the `truncate` method is probably enough as the rest is dependent on the VT implementation. I would simply check that if we perform multiple TRUNCATE all the calls are transmitted to the VT and not only the first one.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 8]
```diff
@@ -88,17 +89,45 @@ public void apply(PartitionUpdate update)
         }
     }
 
+    private abstract static class TruncatableVirtualTable extends AbstractVirtualTable
+    {
+        private boolean isTruncated = false;
```
Comment (blerer) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: Virtual table can be called by multiple threads and that the tread reading the VT might be different from the one performing the truncation.

File: src/java/org/apache/cassandra/cql3/statements/TruncateStatement.java [hunk 9]
```diff
@@ -66,7 +68,10 @@ public ResultMessage execute(QueryState state, QueryOptions options, long queryS
                 throw new InvalidRequestException("Cannot TRUNCATE materialized view directly; must truncate base table instead");
 
             if (metaData.isVirtual())
-                throw new InvalidRequestException("Cannot truncate virtual tables");
+            {
+                executeForVirtualTable(metaData.id);
+                return null;
```
Comment (blerer) [inline src/java/org/apache/cassandra/cql3/statements/TruncateStatement.java]: nit: In this case I would have use a `if-else` to keep a single return. 

I did not remember what we normally return for TRUNCATE statements and I had to check that it was matching the other one.

Comment (azotcsit) [reply src/java/org/apache/cassandra/cql3/statements/TruncateStatement.java]: Done.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java (hunk 8 above)
Comment (azotcsit) [reply test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: Your comment is a bit confusing because I'm not sure whether you are proposing to change something or not. I totally agree on the fact that multiple threads can initiate the truncation. However, I see no point for adding any kind of thread safety here because it is just a test implementation. In fact, in tests there are no occasions of concurrent access to this VT.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java (hunk 4 above)
Comment (azotcsit) [reply test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: > Checking that perfoming a TRUNCATE call the truncate method is probably enough as the rest is dependent on the VT implementation. 

That's what I thought too and decided to keep the simpler version.

> I would simply check that if we perform multiple TRUNCATE all the calls are transmitted to the VT and not only the first one.

This statement sounds a bit confusing. I'm not sure where the assumption that "if the first call is transmitted to the VT, the second one may not be" comes from. In fact, it is pretty easy to substitute the flag by a counter and add the corresponding assertion, but I want to ensure it really makes sense. For me such a counting sounds like a testing of completely another logic ("routing part" is deterministic and is not really related to the VTs). Also following that assumption we may want to change other tests, for example, ensure that ``UPDATE`` is applied twice and so on.

Please, let me know your thoughts.

Comment (blerer) [reply test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: I have seen case in the past where people introduced a bug in the code the tests failed to catch it. So, I was wondering if there were some scenarios that we could miss where the test could give us a false positive. I agree with you that the implementation is straight forward today but being paranoiac, I wished to make the test a bit more bullet proof. The counter approach is pretty simple and give us an extra security at low cost.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java (hunk 8 above)
Comment (blerer) [reply test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: Sorry for being confusing. You are right, with the way the test is done today there is no concurrency race.
I have seen in the past some tests where we decided to switch from `execute` to `executeNet` to do some extra testing for authorization for example. In such case the code will become mutltithreaded and people might miss the fact that the test was not intended to support concurrency initially. To prevent such a problem, I would have made the variable `volatile` to be on the safe side.

File: doc/source/new/virtualtables.rst [hunk 10]
```diff
@@ -66,21 +66,21 @@ Virtual Table Limitations
...
-- Cannot create functions in virtual keyspaces
-- Cannot create types in virtual keyspaces
 - Materialized views are not supported on virtual tables
-- Virtual tables don't support ``DELETE`` statements
+- Virtual tables support modifications only if the underlaying implementation allows it
 - Cannot ``CREATE TRIGGER`` against a virtual table
 - Conditional ``BATCH`` statements cannot include mutations for virtual tables
 - Cannot include a virtual table statement in a logged batch
 - Mutations for virtual and regular tables cannot exist in the same batch
+- Cannot alter or drop virtual keyspaces or tables
```
Comment (azotcsit) [inline doc/source/new/virtualtables.rst:79]: I moved all keyspace-related limitations to the end of the list, so it is easier to read.

File: src/java/org/apache/cassandra/cql3/statements/DeleteStatement.java [hunk 11]
```diff
@@ -177,6 +175,8 @@ protected ModificationStatement prepareInternal(TableMetadata metadata,
 
             if (stmt.hasConditions() && !restrictions.hasAllPKColumnsRestrictedByEqualities())
             {
+                checkFalse(stmt.isVirtual(), "DELETE statements must restrict all PRIMARY KEY columns with equality relations");
```
Comment (azotcsit) [inline src/java/org/apache/cassandra/cql3/statements/DeleteStatement.java:178]: This check is to ensure a proper message (without mentioning static columns) is shown when not all PKs are specified with a condition.

File: src/java/org/apache/cassandra/cql3/statements/ModificationStatement.java [hunk 12]
```diff
@@ -266,6 +266,7 @@ public void validate(ClientState state) throws InvalidRequestException
         checkFalse(isCounter() && attrs.isTimestampSet(), "Cannot provide custom timestamp for counter updates");
         checkFalse(isCounter() && attrs.isTimeToLiveSet(), "Cannot provide custom TTL for counter updates");
         checkFalse(isView(), "Cannot directly modify a materialized view");
+        checkFalse(isVirtual() && attrs.isTimestampSet(), "Custom timestamp is not supported by virtual tables");
```
Comment (azotcsit) [inline src/java/org/apache/cassandra/cql3/statements/ModificationStatement.java:269]: This is to ensure custom timestamps are not supported.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 13]
```diff
@@ -249,76 +344,188 @@ public void testQueriesOnTableWithMultiplePks() throws Throwable
...
+        // test that row DELETE without full primary key with equality relation is (currently) rejected
+        assertInvalidMessage("Some partition key parts are missing: pk2",
+                "DELETE FROM test_virtual_ks.vt2 WHERE pk1 = 'pk1' AND c1 = 'c1' AND c2 > 'c2'");
+        assertInvalidMessage("Only EQ and IN relation are supported on the partition key (unless you use the token() function) for DELETE statements",
+                "DELETE FROM test_virtual_ks.vt2 WHERE pk1 = 'pk1' AND pk2 > 'pk2' AND c1 = 'c1' AND c2 > 'c2'");
+        assertInvalidMessage("KEY column \"c2\" cannot be restricted as preceding column \"c1\" is not restricted",
+                "DELETE FROM test_virtual_ks.vt2 WHERE pk1 = 'pk1' AND pk2 = 'pk2' AND c2 > 'c2'");
+        assertInvalidMessage("Clustering column \"c2\" cannot be restricted (preceding column \"c1\" is restricted by a non-EQ relation)",
+                "DELETE FROM test_virtual_ks.vt2 WHERE pk1 = 'pk1' AND pk2 = 'pk2' AND c1 > 'c1' AND c2 > 'c2'");
+        assertInvalidMessage("DELETE statements must restrict all PRIMARY KEY columns with equality relations",
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:614]: This corresponds to the new restriction in `DeleteStatement`.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 14]
```diff
@@ -249,76 +344,188 @@ public void testQueriesOnTableWithMultiplePks() throws Throwable
...
-        // test that deletions are (currently) rejected
-        assertInvalidMessage("Virtual tables don't support DELETE statements",
-                             "DELETE FROM test_virtual_ks.vt2 WHERE key ='pk1'");
+                "BEGIN UNLOGGED BATCH " +
+                        "UPDATE test_virtual_ks.vt2 SET v1 = 1 WHERE pk1 = 'pk1_1' AND pk2 = 'pk2' AND c1 = 'c1' AND c2 = 'c2';" +
+                        "UPDATE %s                  SET v1 = 2 WHERE pk1 = 'pk1_2' AND pk2 = 'pk2' AND c1 = 'c1' AND c2 = 'c2';" +
+                        "UPDATE test_virtual_ks.vt2 SET v1 = 3 WHERE pk1 = 'pk1_3' AND pk2 = 'pk2' AND c1 = 'c1' AND c2 = 'c2';" +
+                        "APPLY BATCH");
+
+        // test that TIMESTAMP is (currently) rejected with INSERT and UPDATE
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:581]: This corresponds to custom timestamps restriction.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 15]
```diff
@@ -53,52 +57,143 @@
...
-        public void apply(PartitionUpdate update)
+        protected void applyPartitionDelete(Object[] partitionKeyColumnValues)
+        {
+            String pk1 = (String) partitionKeyColumnValues[0];
+            String pk2 = (String) partitionKeyColumnValues[1];
+            backingMap.remove(Pair.create(pk1, pk2));
+        }
+
+        @Override
+        protected void applyRangeTombstone(Object[] partitionKeyColumnValues,
```
Comment (azotcsit) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:99]: I believe for most writable VTs it won't be required to support range deletes. However, it is possible as it is demonstrated in this method.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+            applyColumnUpdate(extractPartitionKeyColumnValues(partitionKey),
+                    extractClusteringColumnValues(clusteringColumns),
+                    extractColumnName(cell),
+                    extractColumnValue(cell));
+        }
+
+        protected abstract void applyColumnUpdate(Object[] partitionKeyColumnValues, Object[] clusteringColumnValues,
+                                                  String columnName, Object columnValue);
+
+        protected <V> void filterStringKeySortedMap(SortedMap<String, V> clusteringColumnsMap,
```
Comment (azotcsit) [inline src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: It is just a utility method that can be re-used in the child classes.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+
+        if (update.deletionInfo().isLive())
+            update.forEach(row ->
+            {
+                Clustering<?> clusteringColumns = row.clustering();
+
+                if (row.deletion().isLive())
+                    row.forEach(columnMetadata ->
+                    {
+                        if (columnMetadata.column().isComplex())
```
Comment (azotcsit) [inline src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: Currently deletes for complex cell types are not supported. I'd like to take care of that in a separate PR when I work on VTs for auth caches.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+import org.apache.cassandra.db.marshal.CompositeType;
+import org.apache.cassandra.db.partitions.PartitionUpdate;
+import org.apache.cassandra.db.rows.Cell;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.schema.TableMetadata;
+
+/**
+ * An abstract virtual table implementation that builds the resultset on demand and allows fine-grained source modification.
+ */
+public abstract class AbstractWritableVirtualTable extends AbstractVirtualTable
```
Comment (blerer) [inline src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: I wonder if we should not collapse the hierachy and merge `AbstractWritableVirtualTable `and `SimpleWritableVirtualTable `together. It would be easy to separate them later on if needed. Having multiple sub-classes that are not used is often a bit confusing and make things harder to understand for no real reason. The `DataSet `hierachy suffer from the same problem, in my opinion, there is an `AbstractDataSet` but it has only a single sub-class. What is your opinion?

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 17 above)
Comment (blerer) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: It is not necessarily a delete. It might also be an update. For now I do not think that we need to support it.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+    {
+        DecoratedKey partitionKey = update.partitionKey();
+
+        if (update.deletionInfo().isLive())
+            update.forEach(row ->
+            {
+                Clustering<?> clusteringColumns = row.clustering();
+
+                if (row.deletion().isLive())
+                    row.forEach(columnMetadata ->
```
Comment (blerer) [inline src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: The variable name `columnMetadata`  is confusing  as it is not a `ColumnMetadata` object but a `ColumnData` one.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 18 above)
Comment (blerer) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: I would rename `AbstractWritableVirtualTable` to `AbstractMutableVirtualTable`, I imagine that we might have some virtual tables that only support deletion but not update (it might be the case for the caches virtual tables).

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+            Object endClusteringColumnValue = endClusteringColumns.isBottom()
+                    ? null : endClusteringColumnValues[endClusteringColumnValues.length - 1];
+            boolean isEndClusteringColumnInclusive = endClusteringColumns.isInclusive();
+
+            applyRangeTombstone(extractPartitionKeyColumnValues(partitionKey),
+                    clusteringColumnValuesPrefix,
+                    startClusteringColumnValue,
+                    isStartClusteringColumnInclusive,
+                    endClusteringColumnValue,
+                    isEndClusteringColumnInclusive);
```
Comment (blerer) [inline src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: Rather than having `startClusteringColumnValue`, `isStartClusteringColumnInclusive`, `endClusteringColumnValue` and `isEndClusteringColumnInclusive` would it not be possible to use a `com.google.common.collect.Range` object? I think that it could help to simplify the code.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+        }
+
+        private String extractColumnName(Cell<?> cell)
+        {
+            return cell.column().name.toCQLString();
+        }
+
+        private Object extractColumnValue(Cell<?> cell)
+        {
+            return metadata.getColumn(cell.column().name).cellValueType().compose(cell.buffer());
```
Comment (blerer) [inline src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: It can be replaced by `cell.column().cellValueType().compose(cell.buffer())` if I am not mistaken.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+    public void apply(PartitionUpdate update)
+    {
+        DecoratedKey partitionKey = update.partitionKey();
+
+        if (update.deletionInfo().isLive())
+            update.forEach(row ->
+            {
+                Clustering<?> clusteringColumns = row.clustering();
+
+                if (row.deletion().isLive())
```
Comment (blerer) [inline src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: For large block like that is it is probably less confusing and less error prone to use parentheses.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 20 above)
Comment (blerer) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: The reasons for extractin the prefix are not fully clear to me? Could you explain me your reasoning?

Comment (blerer) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: The reasons for extracting the prefix are not fully clear to me? Could you explain me your reasoning?

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 23]
```diff
@@ -249,76 +344,188 @@ public void testQueriesOnTableWithMultiplePks() throws Throwable
...
+        // update a single columns with INSERT
+        execute("INSERT INTO test_virtual_ks.vt2 (pk1, pk2, c1, c2, v2) VALUES ('pk1_1', 'pk2_1', 'c1_1', 'c2_2', 22)");
+        assertRows(execute("SELECT * FROM test_virtual_ks.vt2 WHERE pk1 = 'pk1_1' AND pk2 = 'pk2_1' AND c1 = 'c1_1' AND c2 = 'c2_2'"),
+                   row("pk1_1", "pk2_1", "c1_1", "c2_2", 2, 22L));
+
+        // update multiple columns with INSERT
+        execute("INSERT INTO test_virtual_ks.vt2 (pk1, pk2, c1, c2, v1, v2) VALUES ('pk1_1', 'pk2_1', 'c1_1', 'c2_2', 222, 222)");
+        assertRows(execute("SELECT * FROM test_virtual_ks.vt2 WHERE pk1 = 'pk1_1' AND pk2 = 'pk2_1' AND c1 = 'c1_1' AND c2 = 'c2_2'"),
+                row("pk1_1", "pk2_1", "c1_1", "c2_2", 222, 222L));
+
```
Comment (blerer) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:395]: We should also check for updates using `IN` operators

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 24]
```diff
@@ -249,76 +344,188 @@ public void testQueriesOnTableWithMultiplePks() throws Throwable
...
+        // delete a single row
+        execute("DELETE FROM test_virtual_ks.vt2 WHERE pk1 = 'pk1_2' AND pk2 = 'pk2_1' AND c1 = 'c1_1' AND c2 = 'c2_5'");
+        assertRows(execute("SELECT * FROM test_virtual_ks.vt2"),
+                row("pk1_2", "pk2_1", "c1_1", "c2_6", 6, 6L));
+
+        // delete a single column
+        execute("DELETE v1 FROM test_virtual_ks.vt2 WHERE pk1 = 'pk1_2' AND pk2 = 'pk2_1' AND c1 = 'c1_1' AND c2 = 'c2_6'");
+        assertRows(execute("SELECT * FROM test_virtual_ks.vt2"),
+                row("pk1_2", "pk2_1", "c1_1", "c2_6", null, 6L));
```
Comment (blerer) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java]: We should also check for deletion using `IN` operators

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java [hunk 25]
```diff
@@ -0,0 +1,287 @@
...
+                update.deletionInfo()
+                        .rangeIterator(false)
+                        .forEachRemaining(rt -> applyRangeTombstone(partitionKey, rt.deletedSlice()));
+
+            if (!update.deletionInfo().getPartitionDeletion().isLive())
+                applyPartitionDelete(partitionKey);
+        }
+    }
+
+    protected void applyPartitionDelete(DecoratedKey partitionKey)
```
Comment (blerer) [inline src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: Rather than the verb `Delete` I would use the noun `Deletion` in the apply method names.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 19 above)
Comment (azotcsit) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: Good catch! Fixed.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 21 above)
Comment (azotcsit) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: That works and looks clearer, thanks! Done.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 22 above)
Comment (azotcsit) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: My personal preference is to use parentheses for every condition (regardless it is large or small), but based on previous PRs experience I though there is a general preference to not use parentheses if condition body does not contain two statements.

Added parentheses.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 25 above)
Comment (azotcsit) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: In fact I assumed it to be a noun, like "a delete" similarly to "an update". However, I feel `Deletion` makes more sense because this term is used in other classes and sounds more coherent to the project's vocabulary.

Fixed.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 18 above)
Comment (azotcsit) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: +1 on rename - done.

I was thinking about the same. In fact, initially I had a single class, but then decided to follow the structure of `DataSet` and split the classes. Looks like we're not tied to `DataSet` structure, so +1 on collapsing the classes. Done.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 17 above)
Comment (azotcsit) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: You're right! Fixed the error message.

File: src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java (hunk 20 above)
Comment (azotcsit) [reply src/java/org/apache/cassandra/db/virtual/AbstractWritableVirtualTable.java]: I like `Range` idea! Done.

Regarding the prefix, the idea is like that:
All clustering columns except the last level should have equality condition and range comparison is not applicable to them. For example: `c1='c1_1' AND c2='c2_1' AND c3>'c3_1'` it is impossible to specify `c1>='c1_1' AND c2='c2_1' AND c3>'c3_1'` or `c1='c1_1' AND c2<'c2_1' AND c3>'c3_1'`. So basically prefix contains value for all "first" clustering columns. For this example it would be ["c1_1", "c2_1"].

It was mentioned in the comment for the prefix calculation. Now I moved these details to the javadoc for `applyRangeTombstone`.

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java (hunk 23 above)
Comment (azotcsit) [reply test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:395]: Good catch! I added a few test case with `IN` operator.

File: src/java/org/apache/cassandra/db/virtual/AbstractMutableVirtualTable.java [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    private Comparable<?>[] extractClusteringColumnValues(ClusteringPrefix<?> clusteringColumns)
+    {
+        // clusteringColumns.size() may be less than metadata.clusteringColumns().size() since not all clustering
+        // columns have to be always specified
+        Comparable<?>[] clusteringColumnValues = new Comparable<?>[clusteringColumns.size()];
+        for (int i = 0; i < clusteringColumnValues.length; i++)
+        {
+            // VTs do not support UDTs, so casting is safe here because clustering columns of simple types are always comparable
```
Comment (azotcsit) [inline src/java/org/apache/cassandra/db/virtual/AbstractMutableVirtualTable.java]: any concerns on this?

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java [hunk 27]
```diff
@@ -53,52 +57,141 @@
     private static final String VT2_NAME = "vt2";
     private static final String VT3_NAME = "vt3";
 
-    private static class WritableVirtualTable extends AbstractVirtualTable
+    private static class MutableVirtualTable extends AbstractMutableVirtualTable
     {
-        private final ColumnMetadata valueColumn;
-        private final Map<String, Integer> backingMap = new HashMap<>();
+        // <pk1, pk2> -> c1 -> c2 -> <v1, v2>
+        private final Map<Pair<String, String>, SortedMap<String, SortedMap<String, Pair<Integer, Long>>>> backingMap = new ConcurrentHashMap<>();
```
Comment (clohfink) [inline test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:63]: NP: this is really hard to follow, at this depth of generics might be worth having a POJO to wrap part of it

**Review by clohfink — APPROVED** [review]
+1, ignorable NP

File: src/java/org/apache/cassandra/cql3/statements/DeleteStatement.java [hunk 28]
```diff
@@ -145,8 +145,6 @@ protected ModificationStatement prepareInternal(TableMetadata metadata,
                                                         Conditions conditions,
                                                         Attributes attrs)
         {
-            checkFalse(metadata.isVirtual(), "Virtual tables don't support DELETE statements");
```
Comment (clohfink) [inline src/java/org/apache/cassandra/cql3/statements/DeleteStatement.java:148]: I like how little it  took 😄

File: test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java (hunk 27 above)
Comment (azotcsit) [reply test/unit/org/apache/cassandra/cql3/validation/entities/VirtualTableTest.java:63]: I agree that it looks cumbersome. As a way to make it a bit clearer, I added a comment above explaining the structure of `backingMap`. Originally I thought about introducing some POJOs, but finally decided that it would be an overkill for a test implementation. If it does not seem to be a big concern for you I'd keep it as is.

File: src/java/org/apache/cassandra/cql3/statements/DeleteStatement.java (hunk 28 above)
Comment (azotcsit) [reply src/java/org/apache/cassandra/cql3/statements/DeleteStatement.java:148]: Yeah! The most complicated part was just to cover everything with tests :smile:

File: src/java/org/apache/cassandra/db/virtual/AbstractMutableVirtualTable.java (hunk 26 above)
Comment (clohfink) [reply src/java/org/apache/cassandra/db/virtual/AbstractMutableVirtualTable.java]: you can include a precondition check to throw a more formatted error incase supported in future

Comment (azotcsit) [reply src/java/org/apache/cassandra/db/virtual/AbstractMutableVirtualTable.java]: Ok, makes sense. Done.

**Comment by azotcsit** [comment]
Was merged with some fixes and enhancements made by @blerer as f7c71f65c000c2c3ef7df1b034b8fdd822a396d8.
